tests/subsys/bluetooth/mesh/scheduler_model/message_validity/src/main.c

```diff
@@ -0,0 +1,346 @@
+{
+	return 0;
+}
+
+static int action_set(const struct bt_mesh_schedule_entry *test_action,
+		       struct bt_mesh_model *sched_mod, bool ack)
+{
+	BT_MESH_MODEL_BUF_DEFINE(
+		buf, ack ? BT_MESH_SCHEDULER_OP_ACTION_SET : BT_MESH_SCHEDULER_OP_ACTION_SET_UNACK,
+		BT_MESH_SCHEDULER_MSG_LEN_ACTION_SET);
+
+	scheduler_action_pack(&buf, 0, test_action);
+
+	return _bt_mesh_scheduler_setup_srv_op[ack ? 0 : 1].func(sched_mod, NULL, &buf);
+}
+
+static int current_register_get(struct bt_mesh_model *sched_mod)
+{
+	BT_MESH_MODEL_BUF_DEFINE(buf, BT_MESH_SCHEDULER_OP_GET,
+				 BT_MESH_SCHEDULER_MSG_LEN_GET);
+
+	return _bt_mesh_scheduler_srv_op[0].func(sched_mod, NULL, &buf);
+}
+
+static int action_get(struct bt_mesh_model *sched_mod,
+				uint8_t idx)
+{
+	BT_MESH_MODEL_BUF_DEFINE(buf, BT_MESH_SCHEDULER_OP_ACTION_GET,
+				 BT_MESH_SCHEDULER_MSG_LEN_ACTION_GET);
+
+	net_buf_simple_add_u8(&buf, idx);
+
+	return _bt_mesh_scheduler_srv_op[1].func(sched_mod, NULL, &buf);
+}
+
+
+static void setup(void)
+{
+	mod_elem_idx_prep();
+
+	/* 1st of Jan 2010 */
+	start_tm.tm_year = 110;
+	start_tm.tm_mon = 0;
+	start_tm.tm_mday = 1;
+	start_tm.tm_wday = 5;
+	start_tm.tm_hour = 0;
+	start_tm.tm_min = 0;
+	start_tm.tm_sec = 0;
+
+	zassert_not_null(_bt_mesh_scheduler_srv_cb.init,
+			 "Init cb is null");
+
+	_bt_mesh_scheduler_srv_cb.init(sched_mod_elem);
+}
+
+static void teardown(void)
+{
+	zassert_not_null(_bt_mesh_scheduler_srv_cb.reset,
+			 "Reset cb is null");
+
+	_bt_mesh_scheduler_srv_cb.reset(sched_mod_elem);
+}
+
+
+static void test_set_prohibited(void)
+{
+	/** Attempt to send a SET / SET_UNACK message with a prohibited schedule time.
+	 * 1. Test that unacked set server message handler returns with error.
+	 * 2. Test that acked set server message handler returns with error.
+	 */
+
+	const struct bt_mesh_schedule_entry test_action = {
+		.year = 0x65, /* prohibited year */
+		.month = ANY_MONTH,
+		.day = 0,
+		.hour = 24,
+		.minute = 0,
+		.second = 30,
+		.day_of_week = ANY_DAY_OF_WEEK,
+		.action = BT_MESH_SCHEDULER_TURN_ON,
+		.transition_time = model_transition_encode(2500),
+		.scene_number = 1,
+	};
+
+	zassert_equal(action_set(&test_action, sched_mod_elem, false), -EINVAL);
+	zassert_equal(action_set(&test_action, sched_mod_elem, true), -EINVAL);
+}
+
+static void test_set_valid(void)
+{
+	/** Send a SET / SET_UNACK message with a valid schedule time.
+	 * 1. Test that unacked set message handler returns with success.
+	 * 2. Test that acked set message handler returns with success.
+	 */
+
+	const struct bt_mesh_schedule_entry test_action = {
+		.year = 10,
+		.month = ANY_MONTH,
+		.day = 0,
+		.hour = 24,
+		.minute = 0,
+		.second = 30,
+		.day_of_week = ANY_DAY_OF_WEEK,
+		.action = BT_MESH_SCHEDULER_TURN_ON,
+		.transition_time = model_transition_encode(2500),
+		.scene_number = 1,
+	};
+
+	zassert_false(action_set(&test_action, sched_mod_elem, false));
+	zassert_false(action_set(&test_action, sched_mod_elem, true));
```

Comment:
The same for valid tests. Test should give assurance if we provide model correct data as income we get back correct data as outcome. 
Unit tests should assume tested unit (scheduler model) has unknown implementation. However, we know how black box should behave itself. If provide some input and get back expected output that means unknown implementation behaves itself correctly and we Ok with it.   

Your test checks the only immediate model reaction aka returned value. This is Ok but it should check reaction on its outcome too.

Reply:
Just to get some more clarity on this: do you want me to actually verify that the set action is executed at the correct time? I feel like this is already covered by Anders action planning test. At least for the SET-operations. Or do you just want to verify that the server sends a (correct) reply in the case of acknowledged set operation?

Reply:
Agree, the first one is covered by Anders test.
I meant the second one. Additionally, to this, it is worth to check that the appropriate callback(action_set_cb) is called.
And the case where status is sent once and two times (1- as a reply\status publish, 2 - as status publishing).